File: benchmarks/fixtures/webapp_csharp/Database/DatabaseConnection.cs
namespace Webapp.Database;

using System.Collections.Generic;
using Webapp.Util;

/// <summary>
/// Represents a single database connection with query execution support.
/// </summary>
public class DatabaseConnection
{
    private static readonly Logger Log = Logger.GetLogger("database.connection");
    private readonly ConnectionPool _pool;

    public DatabaseConnection(string host, int port, string database)
    {
        Log.Info($"Creating database connection: {host}:{port}/{database}");
        _pool = new ConnectionPool(10);
        Log.Info("Database connection established");
    }

    /// <summary>
    /// Execute a query and return results.
    /// </summary>
    public List<Dictionary<string, object>> ExecuteQuery(string query)
    {
        Log.Info($"Executing query: {query}");
        ConnectionHandle handle = _pool.GetConnection();
        Log.Debug($"Query executed on connection #{handle.Id}");
        _pool.ReleaseConnection(handle);
        return new List<Dictionary<string, object>>();
    }

    public string Insert(string table, Dictionary<string, object> data)
    {
        Log.Info($"Insert into table: {table}");
        ExecuteQuery($"INSERT INTO {table} VALUES (?)");
        return "generated_id";
    }
}

/// <summary>
/// Generic repository over a table (D3: generics → clean base name Repository).
/// </summary>
public class Repository<T> where T : class
{
    private readonly DatabaseConnection _db;

    public Repository(DatabaseConnection db)
    {
        _db = db;
    }

    public void Save(T entity)
    {
        _db.ExecuteQuery("INSERT");
    }
}

File: benchmarks/fixtures/webapp_cpp/database/database_connection.cpp
#include "database/database_connection.h"

#include <string>
#include <vector>

#include "database/connection_pool.h"
#include "util/logger.h"

namespace webapp {
namespace database {

DatabaseConnection::DatabaseConnection(const std::string& host, int port,
                                       const std::string& database)
    : log_(util::Logger::get_logger("database.connection")),
      pool_(10),
      host_(host),
      port_(port),
      database_(database) {
    log_.info("Creating database connection: " + host_ + ":" + std::to_string(port_) + "/" +
              database_);
    log_.info("Database connection established");
}

/// Execute a query and return its rows.
std::vector<Row> DatabaseConnection::execute_query(const std::string& query) {
    log_.info("Executing query: " + query);
    ConnectionHandle handle = pool_.get_connection();
    log_.debug("Query executed on connection #" + std::to_string(handle.id()));
    pool_.release_connection(handle);
    return std::vector<Row>();
}

/// Insert a row into a table, returning the generated id.
std::string DatabaseConnection::insert(const std::string& table, const Row& data) {
    log_.info("Insert into table: " + table + " (" + std::to_string(data.size()) + " columns)");
    execute_query("INSERT INTO " + table + " VALUES (?)");
    return "generated_id";
}

const std::string& DatabaseConnection::database() const {
    return database_;
}

}  // namespace database
}  // namespace webapp
